### PR TITLE
Circuit diagrams for unitary quantum kernels in LaTeX

### PR DESCRIPTION
**Pull request by freifrauvonbleifrei**
Add LaTeX circuit diagrams to the cudaq draw() function.

### Description

this PR addresses https://github.com/NVIDIA/cuda-quantum/issues/1638

### CODE REVIEW

**Comment by freifrauvonbleifrei** [comment]
recheck

**Comment by freifrauvonbleifrei** [comment]
I am currently building the latex reference output for testing, to match this string output: 
https://github.com/NVIDIA/cuda-quantum/blob/1c76ab2e459eba86e8aa147acf7a23963cf21abd/unittests/integration/draw_tester.cpp#L66-L86

This is the current draft:
```
\documentclass{minimal}
\usepackage{quantikz}
\begin{document}
\begin{quantikz}
  \lstick{$q_0$} & \gate{h} & \ctrl{1} & \ctrl{2} & \ctrl{1} & \gate{y}  & \gate{r_1(3.142)} & \qw     & \swap{2} & \qw      & \swap{1} & \swap{2} & \qw      & \targX{} & \ctrl{2} & \swap{3} & \swap{3} & \gate{r_x(2.718)} & \gate{\text{sdg}} & \gate{r_x(-2.718)} & \qw \\
  \lstick{$q_1$} & \gate{h} & \gate{x} & \ctrl{1} & \gate{y} & \ctrl{-1} & \gate{\text{tdg}} & \qw     & \qw      & \swap{1} & \targX{} & \qw      & \swap{1} & \targX{} & \targX{} & \ctrl{2} & \ctrl{2} & \gate{r_y(3.142)} & \ctrl{-1}         & \gate{r_y(-3.142)} & \qw \\
  \lstick{$q_2$} & \gate{h} & \qw      & \gate{y} & \ctrl{-1}& \ctrl{-2} & \gate{z}          & \gate{s}& \targX{} & \targX{} & \qw      & \targX{} & \targX{} & \qw      & \targX{} & \qw      & \ctrl{-2}& \gate{r_z(-3.142)}& \gate{r_z(3.142)} & \qw                & \qw \\
  \lstick{$q_3$} & \gate{h} & \qw      & \qw      & \qw      & \qw       & \qw               & \qw     & \qw      & \qw      & \qw      & \qw      & \qw      & \ctrl{-3}& \ctrl{-2}& \targX{} & \targX{} & \qw               & \qw               & \qw                & \qw \\
\end{quantikz}
\end{document}
```
which produces this
when run with `latex` and `dvisvgm`.

I think that `.svg` output should be the easiest to integrate with jupyter notebooks.

My svg viewer shows weird triangles for the numerical dots (they aren't there when I run `dvisvgm --no-fonts`, but then all semantics is lost, too). I would just trust that these display issues will be fixed somewhere along the toolchain, and accept them for now :shrug: 

@bettinaheim @amccaskey @boschmitt does this look somewhat like the thing you'd like?

**Comment by freifrauvonbleifrei** [comment]
I think this PR fulfills the requirements stated in issue #1638. 

However, I would like to do some more notebook integration. 
This would mean
* having a Python function that invokes subprocesses to produce the `.dvi` and `.svg` intermediate files, and returns the contents of the `.svg` as string. (Questions: Where does it belong in the python module? What happens in case of failure, e.g. if `latex` or `dvisvgm` are not installed, or files are not writable?)
* wrapping this in a `_repr_svg_` function [that can be understood by IPython](https://ipython.readthedocs.io/en/stable/config/integrating.html#rich-display) . But here is the tricky part: which class should this function belong to? To obtain the trace, both the kernel and its arguments are required.

Opinions are highly appreciated!

**Comment by freifrauvonbleifrei** [comment]
The latest commit lets you display the svg in a jupyter notebook. Here, for instance, I added `displaySVG` to the quantum Fourier transform tutorial:
image
I am not entirely happy about the Python import structure though, but cannot think of a more elegant way to do it right now. Feedback is appreciated!

**Comment by freifrauvonbleifrei** [comment]
And I added a `_repr_svg_` for `PyKernelDecorator`, which will output the svg if the kernel needs no parameters. 
Here in the `quantum_bits.ipynb` tutorial:
image

**Comment by boschmitt** [comment]
Hi @freifrauvonbleifrei, thanks for all the work put into this.

I noticed that you are using C++20 concepts in the code. We also use it, however, you need to provide an alternative so that the code can also compile with C++17. For example:

https://github.com/NVIDIA/cuda-quantum/blob/b522e18b0bfe0c931db3e2ee243dd01752597e19/runtime/cudaq/algorithms/sample.h#L242-L249

**Comment by freifrauvonbleifrei** [comment]
Hi @boschmitt , thank you for having a look. I was not aware that C++17 compatibility would be required, and it is especially surprising to me that other places in the code use concepts unguarded:
https://github.com/NVIDIA/cuda-quantum/blob/6c736fa3ea133cdf1dd86fa612b45c1a01eef566/include/cudaq/Optimizer/Builder/Factory.h#L29
How are these other parts ensured to be compiled with C++20? 

I can add some guards for sure, but I would like to make sure this is what is needed in the project, as readability would really suffer.

**Comment by boschmitt** [comment]
> I was not aware that C++17 compatibility would be required, and it is especially surprising to me that other places in the code use concepts unguarded:
> 
> https://github.com/NVIDIA/cuda-quantum/blob/6c736fa3ea133cdf1dd86fa612b45c1a01eef566/include/cudaq/Optimizer/Builder/Factory.h#L29
> 
> 
> How are these other parts ensured to be compiled with C++20?

I was not sufficiently precise in my observation. The thing is: not all parts of the codebase have the same requirements regarding C++17 support. We need C++17 support on the public headers, i.e., the headers users would import (directly or indirectly) to develop their quantum applications. (In other words, to build the project, one needs a C++20 capable compiler. However, to use it, you only need C++17.)

The code you showed is not part of the public API and is not installed (distributed with CUDAQ). Thus, there is no need to comply with the C++17 requirement.

File: python/tests/display/test_draw.py [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+
+import cudaq
+import numpy as np
+import os
+import pytest
+
+
+@cudaq.kernel
+def bar(qvec: cudaq.qview):
+    # FIXME https://github.com/NVIDIA/cuda-quantum/issues/1734
```
Comment (bettinaheim) [inline python/tests/display/test_draw.py:17]: Thanks for filing the issue, and submitting PR 1756 as well!

**Review by bettinaheim — APPROVED** [review]
Nice! Overall this looks very nice. 
Let's follow up in a separate PR for the C++ 17 support in the public headers, so that we can stick to the Unitary Hack time lines.

**Pull request merged**